### ISSUE TITLE
support vlan aware bridge

### ISSUE THREAD

**Issue by saschalucas** [ganeti/ganeti#1387]
nowadays the linux bridge can be VLAN aware. this would simplify network setup: only one bridge device is needed, instead one bridge per VLAN.

**Comment by atta** [comment]
enable vlan aware bridge, can be done as son as a vlan in adden to the gnt-network and the instance start
```
ip link set brx type bridge vlan_filtering 1
```

**Comment by atta** [comment]
i think i sould be only the net-common file, so someting like:

```
function setup_bridge {
  if [ "$MODE" = "bridged" ]; then
    fix_mac
    ip link set $INTERFACE up
    ip link set $INTERFACE mtu $(</sys/class/net/${LINK}/mtu)

    # Connect the interface to the bridge
    ip link set dev $INTERFACE master $LINK

    # Set up access port
    # From gnt-instance man page vlan should be either .VLAN_ID or VLAN_ID
    ACPORT=${VLAN%%:*}  # remove any trunk info
    if [ -n "$ACPORT" ]; then
      ip link set $LINK type bridge vlan_filtering 1
      bridge vlan add vid $ACPORT pvid untagged dev $INTERFACE
    fi

    # Set up trunk port
    # From gnt-instance man page vlan should be :VLAN_ID[:VLAN_ID2..]
    TRUNKS=${VLAN#.*:}  # remove any access info
    TRUNKS=${TRUNKS#:}  # remove leading ':', if still present
    if [ -n "$TRUNKS" ]; then
      ip link set $LINK type bridge vlan_filtering 1
      for TRUNK in ${TRUNKS//:/ }; do
        bridge vlan add vid $TRUNK untagged dev $LINK
      done
    fi
  fi
}
```

**Comment by saschalucas** [comment]
thanks a lot for pointing out on how this can be done. currently I'm preparing a test setup to figure this out. a view questions pop up:

* should the bridge command operate on the $LINK (the bridge) or $INTERNFACE (the tap)? I think $INTERFACE would be right?
* should vlan_filtering turned on outside of this script? i.e. by the administrator setting up the node network? or do you think this is free from any side effects, regardless of how fancy the node networking might be?
* it seems to me, that the code above does not manage a hybrid port (VLAN=.x:y:z)? I've created an alternative (not yet ready) code to do so:
```
VLANS=( ${VLAN//:/ } )
num_vlans=${#VLANS[@]}
# access port
if [[ ${num_vlans} -eq 1 ]]; then
        bridge vlan add vid ${VLANS#.} dev ${INTERFACE} pvid untagged
elif [[ ${num_vlans} -gt 1 ]]; then
        first_vid="${VLANS[0]}"
        # hybrid port native/pvid
        if [[ "${first_vid:0:1}" == "." ]]; then
                bridge vlan add vid ${first_vid#.} dev ${INTERFACE} pvid untagged
        else
                # first tagged
                bridge vlan add vid ${first_vid} dev ${INTERFACE}
        fi
        # other tagged
        for vid in ${VLANS[@]:1}; do
                bridge vlan add vid ${vid} dev ${INTERFACE}
        done
fi
```

**Comment by saschalucas** [comment]
solved with PR #1395
